## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# What three item categories had the most entries?
#categories = df.category.value_counts()
#categories.head(3)
df.category.value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?
#df[["category", "subcategory"]].value_counts()
df.loc[df["category"] == "consumables", "subcategory"].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
most_entries = df.client_id.value_counts().head(5)
most_entries

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_clients = most_entries.index.to_list()
top_clients

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df.client_id == 33615, "qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df.unit_price * df.qty
df[["unit_price", "qty", "subtotal"]].head(3)

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df['total_weight'] = df.unit_weight * df.qty
def calculate_shipping_price(total_weight):
    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10
df['shipping_price'] = df.total_weight.apply(calculate_shipping_price)
df[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]].head()


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = round((df.subtotal + df.shipping_price) * 1.0925 ,2)
df[["subtotal", "shipping_price", "total_price"]].head()


subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df.unit_cost * df.qty) + df.shipping_price
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  total_weight  shipping_price  total_price  \
0            1  115164.00        787.50         5512.50    131839.08   
1            0     523.95         31.29          312.90       914.26   
2            6     527.28         65.52          458.64      1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [15]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = (df.total_price - df.line_cost) 
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price  total_price  line_cost  \
0  115164.00        787.50         5512.50    131839.08   85597.05   
1     523.95         31.29          312.90       914.26     629.79   
2     527.28         65.52          458.64      1077.12     765.18   
3    1056.18         35.67          356.70      1543.57    1077.35   
4    3902.00        928.60         6500.20     11364.40    8663.60   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
total_orders=[2742071, 2173913, 6128929]
for order in total_orders:
    receipts = round( df.loc[df["order_id"] == order, "total_price"].sum() ,2)
    print (f"Order ID {order}  has a total price of ${receipts}")
#df[["order_id", "total_price"]].head(5)
#df[["Order ID", 'order_id', "has a total price of $", 'total_pice']].head()



Order ID 2742071  has a total price of $152811.89
Order ID 2173913  has a total price of $162388.71
Order ID 6128929  has a total price of $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients
for client in top_clients:
    top_five_total_purchases = round( df.loc[df["client_id"] == client, "total_price"].sum() ,2)
    print (f"Client ID {client}  {top_five_total_purchases}")

Client ID 33615  8377308.52
Client ID 66037  10259514.79
Client ID 46820  9743794.36
Client ID 38378  12906550.87
Client ID 24741  82268892.02


In [22]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue= ('total_price', 'sum'),
    total_profit=('line_profit', 'sum')
)

# Sort by total revenue or total profit to get the top 5 clients
top_clients = summary_df.sort_values(by='total_profit', ascending=False).head(5)

# Display the summary DataFrame
top_clients

total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
78965                    4033302            5340756.88   6.559237e+08   
85931                     644423           10122549.19   2.374170e+08   
38473                     402770            4391003.73   1.835526e+08   
86721                    1010426           64335727.72   1.812663e+08   
24741                     239862            5126448.37   8.226889e+07   

           total_profit  
client_id                
78965      2.473528e+08  
85931      1.092941e+08  
38473      4.335173e+07  
86721      4.297470e+07  
24741      3.657999e+07

In [19]:
# Format the data and rename the columns to names suitable for presentation.
renamed_df = df.rename(columns={"client_id":"Client ID",
                                "qty":"Units", 
                                "shipping_price":"Shipping (millions)",
                                "total_price":"Total Revenue (millions)",
                                "line_cost":"Total Cost (millions)", 
                                "line_profit": "Total Profit (millions)"})
                               
renamed_df.head()
# Define the money columns. 


# Define a function that converts a dollar amount to millions.
def convert_to_millions(dollars):
    return dollars / 1_000_000

   
# Apply the currency_format_millions function to only the money columns. 
renamed_df[['Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']] \
    = renamed_df[['Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']].apply(convert_to_millions)

# Rename the columns to reflect the change in the money format. 
renamed_df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  Client ID  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_cost unit_weight  \
0   8953482  2023-04-28          17        2023  ...    762.71        7.50   
1   8069089  2023-05-19          20        2023  ...     15.09        1.49   
2   1902144  2023-01-29           4        2023  ...      7.86        1.68   
3   9031802  2023-04-25          17        2023  ...     24.85        1.23   
4   1322274  2023-05-28          21        2023  ...    108.17       46.43   
5   1327730  2023-05-02          18        2023  ...    154.44        2.79   
6   8080280  2023-04-19          16        2023  ...    231.95        1.73   
7   8946782  2023-02-14           7        2023  ...     16.97        5.12   
8   9394118  2023-05-13          19        2023  ...     59.00        0.27   
9   9665852  2023-04-14          15        2023  ...     54.15        2.47   

  Units  line_number   subtotal  total_weight  Shipping (millions)  \
0   105            1  115164.00        787.50             0.005513   
1    21            0     523.95         31.29             0.000313   
2    39            6     527.28         65.52             0.000459   
3    29            3    1056.18         35.67             0.000357   
4    20            1    3902.00        928.60             0.006500   
5    78            0   13481.52        217.62             0.001523   
6    56            5   13751.36         96.88             0.000678   
7    22            4     461.78        112.64             0.000788   
8    49            3    4109.14         13.23             0.000132   
9    78            1    5216.64        192.66             0.001349   

   Total Revenue (millions)  Total Cost (millions)  Total Profit (millions)  
0                  0.131839               0.085597                 0.046242  
1                  0.000914               0.000630                 0.000284  
2                  0.001077               0.000765                 0.000312  
3                  0.001544               0.001077                 0.000466  
4                  0.011364               0.008664                 0.002701  
5                  0.016393               0.013570                 0.002823  
6                  0.015764               0.013667                 0.002097  
7                  0.001366               0.001162                 0.000204  
8                  0.004634               0.003023                 

In [20]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_data_df = renamed_df.sort_values("Total Profit (millions)", ascending=False)
sorted_data_df.head()

first      last                                    job  \
21801    Julie  Anderson                    Exhibition designer   
15112   Dennis   Hawkins    Social research officer, government   
25972      Amy   Stewart              Psychologist, educational   
23929  Michele  Mcdonald  Emergency planning/management officer   
4063    Kendra   Garrett                 Pension scheme manager   

                       phone                            email  Client ID  \
21801             5933099462    anderson_better_4704@bell.biz      78965   
15112  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
25972     979.793.4884x57205           stewartamy@francis.com      86721   
23929     435-931-4939x95137    mcdonaldm317@hess-jackson.com      38473   
4063            808.257.2029  garrettkendra8479@jefferson.com      24741   

       order_id  order_date  order_week  order_year  ... unit_cost  \
21801   3833568  2023-04-29          17        2023  ...    100.11   
15112   7263857  2023-03-05           9        2023  ...    183.84   
25972   6057452  2023-03-21          12        2023  ...     73.05   
23929   4594154  2023-03-02           9        2023  ...    568.06   
4063    5167984  2023-05-08          19        2023  ...    247.94   

      unit_weight    Units  line_number      subtotal  total_weight  \
21801        0.09  3958244            1  5.844743e+08     356241.96   
15112        2.21   637318            3  2.061214e+08    1408472.78   
25972        9.22   977121            3  9.810295e+07    9009055.62   
23929        2.25   219645            5  1.482472e+08     494201.25   
4063         1.86   122166            7  5.635518e+07     227228.76   

       Shipping (millions)  Total Revenue (millions)  Total Cost (millions)  \
21801             2.493694                641.262543             398.753501   
15112             9.859309                235.958911             127.023851   
25972            63.063389                176.074224             134.442078   
23929             3.459409                165.739466             128.230947   
4063              1.590601                 63.305762              31.880439   

       Total Profit (millions)  
21801               242.509042  
15112               108.935061  
25972                41.632146  
23929                37.508519  
4063                 31.425322  

[5 rows x 24 columns]